designs/2018-simplified-package-loading/README.md

```diff
@@ -0,0 +1,87 @@
+* The current behavior is very complex, amplifying user confusion and making it difficult to maintain and enhance config-loading behavior.
+
+### Appendix: Current package-loading behavior
+
+This section describes the *current* package-loading behavior (before implementing this RFC) for various types of packages, and notes any oddities for posterity/compatibility analysis.
+
+* Plugins are loaded by calling `require(pluginName)` relative to `lib/config/plugins.js`. Since plugin names always have `eslint-plugin-` prepended, they can't contain relative paths, so they are effectively always loaded from the root of the `eslint` package.
+* Parsers and shareable configs are loaded in similar ways:
+    * If a shareable config or parser is referenced from a config file which is inside the same `node_modules` folder as the `eslint` package, ESLint attempts to load it from the sibling `node_modules` folder of the config where the parser is referenced. If it's not found there, then it's effectively loaded from the `eslint` package rather than using Node's regular `node_module` cascade (where Node looks for packages `node_modules` folders in parent directories).
+    * If referenced from a config file which is *not* inside the same `node_modules` folder as the `eslint` package (including config files in the end-user's project), extended configs and parsers are loaded relative to the `eslint` package.
+    * If referenced outside a config file (e.g. on the command line), parsers are loaded relative to the `lib/linter.js` within the `eslint` package.
+
+The current behavior gives the impression that ESLint's config-loading logic evolved somewhat haphazardly in response to new use cases, while still adding patches to maintain compatibility with all existing use cases. Unfortunately, this led to substantial complexity in implementation and behavior, making it difficult for users and maintainers to understand how ESLint was loading packages, and leading to subtle design bugs like [eslint/eslint#10125](https://github.com/eslint/eslint/issues/10125). This RFC should be seen as a substantial simplification to the config-loading model which leaves more behavior to Node's module-resolution algorithm rather than implementing path resolution from scratch, and provides a more consistent mental model for future changes.
+
+## Detailed Design
+
+### Changes to package-loading
+
+With the new design, ESLint will load all plugins and formatters relative to the *project root*. By default, the project root is simply the CWD of the running process. Integrations can configure a specific project root using the already-existing `cwd` option in `CLIEngine`. ESLint will load all other third-party packages (e.g. shareable configs and parsers) relative to the config file where they're referenced.
```

Comment:
I've thought of the project root as the location where the config file is that contains `root: true`. In cases where this isn't found, I think using the CWD makes sense, but in cases where there is a config file with `root: true`, wouldn't it make more sense to resolve from that directory? That would allow you to run `eslint` in any subdirectory and still have it work as expected. (In any case, I think it makes sense to separate the concept of project root from CWD.)

Also, here you use the term "project root" but later on you use the term "package root." It might be helpful to setup those definitions formally in this RFC.

Reply:
to be honest (and i realize this would be a separate RFC), it'd be much nicer to invert that behavior in eslint - ie, make every config be a "root", and require explicit opting in to "extending upwards". That would make it much less ambiguous and avoid needing to rely on the cwd.

Reply:
I'd be okay with defining it as the location of a `root: true` config. However, it's worth noting that packages would still load as expected even if `eslint` is run from a subdirectory, because plugins would still be resolvable from the subdirectory. (For example, if I run `require('eslint-plugin-node')` from `./packages/default/eslint-config-eslint` in the ESLint repo, it will resolve to the plugin in `./node_modules/eslint-plugin-node`.) This seems like it could also cause some confusion if there are multiple config files with `root: true` active simultaneously for two different files being linted, which leads me to slightly favor using the CWD. I don't feel strongly about this, so I'd be open to further discussion/changing it if needed.

Reply:
I see, thanks for explaining. I might opt to just remove the term "project root" from the description in that case and just stick with CWD.

Reply:
See https://github.com/eslint/rfcs/pull/7#issuecomment-454526068